src/parsers/vmap.js

```diff
@@ -0,0 +1,114 @@
+import { Log } from '@clappr/core'
+import AdBreak from '../entities/ad-break'
+import { hmsToMilliseconds, xml2json } from '../utils'
+
+export default class VMAPManager {
+  /**
+   * Request VMAP XML.
+   * @param {String} URL ad server URL to request VMAP file.
+   * @returns {Promise} Promise resolved with VMAP parsed to plain text or one error.
+   */
+  request(url) {
+    return fetch(url)
+      .then(response => {
+        if (!response.ok) throw new Error('Network response was not ok')
+        return response.text()
+      })
+      .catch(error => Promise.reject(error))
+  }
+
+  /**
+   * Parses VMAP XML in plain text to JSON.
+   * @param {String} responseInPlainText Response of the fetch triggered for consuming VMAP.
+   * @returns {Promise} Promise resolved with raw AdBreaks list.
+   */
+  process(responseInPlainText) {
+    const parser = new DOMParser()
+    const xml = parser.parseFromString(responseInPlainText, 'application/xml')
+    const rawAdBreaks = xml2json(xml.documentElement)
+
+    return Promise.resolve(rawAdBreaks)
+  }
+
+  /**
+   * Receives one VMAP formatted response and return a list of AdBreaks instances.
+   * @param {Object} rawData VMAP formatted on JSON schema.
+   * @returns {Promises} An Array of all AdBreaks instances created or one error.
+   */
+  filterRawData(rawData) {
+    return rawData['@version'] && rawData['vmap:AdBreak']
+      ? this._filterStandardRawData(rawData)
+      : this._filterCustomRawData(rawData)
+  }
+
+  _filterStandardRawData(rawData) {
+    const adBreaks = []
+
+    try {
+      rawData['vmap:AdBreak'].forEach(item => {
+        let timeOffset = item['@timeOffset']
+        let category = ''
+
+        if (timeOffset === 'start') {
+          category = 'preroll'
+          timeOffset = null
+        } else if (timeOffset === 'end') {
+          category = 'postroll'
+          timeOffset = null
+        } else {
+          category = 'midroll'
+        }
+
+        const adBreak = this.createAdBreak({
+          category,
+          adTag: item['vmap:AdSource']['vmap:AdTagURI'],
+          timeOffset,
+        })
+
+        adBreaks.push(adBreak)
+      })
+    } catch (error) {
+      Log.error(this.name, 'Build AdBreak process fail', error)
+      return Promise.reject(error)
+    }
+
+    return Promise.all(adBreaks)
```

Comment:
I think `Promise.resolve(adBreaks)` would be clearer here. `Promise.all` gives the idea that each ad break is an asynchronous operation and we are waiting for them to fulfill here. Since ad breaks are plain objects, `Promise.resolve` works just fine.

Reply:
Totally agreed!